localstack/services/s3/s3_listener.py

```diff
@@ -1307,5 +1320,91 @@ def return_response(self, method, path, data, headers, response, request_handler
                 response.headers['Content-Encoding'] = 'gzip'
 
 
+def authenticate_presign_url(method, path, data=None, headers={}):
+    # ignored_headers conatins headers which don't get involved in signature calculations process
+    # these headers are being sent by the localstack by default.
+    ignored_headers = [
+        'Remote-Addr', 'Host', 'User-Agent', 'Accept-Encoding',
+        'Accept', 'Connection', 'Origin',
+        'X-Forwarded-For', 'x-localstack-edge', 'Authorization'
+    ]
+
+    sign_headers = []
+    url = 'http://localhost:4566{}'.format(path)
+    parsed = urlparse.urlparse(url)
+    query_params = parse_qs(parsed.query)
+    AWS_ACCESS_KEY = 'temp'
+    AWS_ACCESS_SECRET_KEY = 'temp'
+
+    # Checking required parameters are present in url or not
+    if 'Signature' not in query_params or 'Expires' not in query_params or 'AWSAccessKeyId' not in query_params:
+        return requests_error_response_xml_presign_url_auth(
+            code=403,
+            message='Query-string authentication requires the Signature, Expires and AWSAccessKeyId parameters',
+            code_string='AccessDenied'
+        )
+
+    # Fetching headers which has been sent to the requets
+    for header in headers:
+        key = header[0]
+        if key not in ignored_headers:
+            sign_headers.append(header)
+
+    # Request's headers are more essentials than the query parameters in the requets.
+    # Different values of header in the header of the request and in the query paramter of the requets url
+    # will fail the signature calulation. As per the AWS behaviour
+    presign_query_params = ['Signature', 'Expires', 'AWSAccessKeyId']
+    if len(query_params) > 2:
+        for key in query_params:
+            if key not in presign_query_params:
+                if key.lower() not in (header[0].lower() for header in headers):
```

Comment:
We can expect `headers` to be a case-insensitive dict here, so we should be able to simply use this:
```
if key not in headers:
    ...
```

Reply:
I am comparing each headers into lowercase because it might be possible that user add headers in first-letter capitals. Also presigned headers are going to be present in the presigned URL (like `'Content-Type': 'application/json'` is going to be present in url as` '&content-type=application/json'`. We are giving priority to the request header's values not those headers who are in query parameter. So whatever we are getting in request's headers we will consider that while calculating the signature (as AWS works). Also there is a case that we might add some headers while generating presign url but not when we actually we hit that url. In that case we are checking that if key is present or not in headers, if it is then we are ignoring those headers (as per request header's value are more important than the query parameter header's value). So to making that comparison we need to make it case-insensitive.